samcli/lib/docker/log_streamer.py

```diff
@@ -0,0 +1,93 @@
+"""
+Log streaming utilities when streaming logs from Docker
+"""
+import os
+from typing import Dict
+
+import docker
+
+from samcli.lib.package.stream_cursor_utils import (
+    CursorUpFormatter,
+    CursorDownFormatter,
+    CursorLeftFormatter,
+    ClearLineFormatter,
+)
+from samcli.lib.utils.stream_writer import StreamWriter
+
+
+class LogStreamError(Exception):
+    def __init__(self, msg: str) -> None:
+        Exception.__init__(self, msg)
+
+
+class LogStreamer:
+    def __init__(self, stream: StreamWriter):
+        self._stream = stream
+        self._cursor_up_formatter = CursorUpFormatter()
+        self._cursor_down_formatter = CursorDownFormatter()
+        self._cursor_left_formatter = CursorLeftFormatter()
+        self._cursor_clear_formatter = ClearLineFormatter()
+
+    def stream_progress(self, logs: docker.APIClient.logs):
+        """
+        Stream progress from docker push logs and move the cursor based on the log id.
+        :param logs: generator from docker_clent.APIClient.logs
```

Comment:
did we consider using https://github.com/tqdm/tqdm or another well-known python library?

Reply:
This is to mimic how docker cli showcases the logs, so didn't want to use another library or abstraction.